contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,144 @@
+        let inner_type = &type_path.path.segments.iter()
+            .map(|ps| ps.ident.clone())
+            .collect::<Vec<Ident>>();
+
+        let database_name = &invocation.db_name;
+        let handler_type = &invocation.type_name;
+        let pool_type = Ident::new(&format!("{}Pool", handler_type), Span::call_site());
+
+        let tokens = quote! {
+            use ::std::ops::*;
+            use ::rocket::http::*;
+            use ::rocket::request::*;
+            use ::rocket::*;
+            use ::rocket::fairing::*;
+            use ::rocket_contrib::*;
+            use ::rocket_contrib::databases::*;
+            pub struct #handler_type(r2d2::PooledConnection<<#(#inner_type)::* as Poolable>::Manager>);
+            pub struct #pool_type(r2d2::Pool<<#(#inner_type)::* as Poolable>::Manager>);
+
+            impl #handler_type {
+                pub fn pool(config: DatabaseConfig) -> Result<#pool_type, <#(#inner_type)::* as Poolable>::Error> {
+                    <#(#inner_type)::* as Poolable>::manager(&config)
+                        .and_then(|manager| {
+                             match r2d2::Pool::builder().max_size(config.pool_size).build(manager) {
+                                 Ok(pool) => Ok(pool),
+                                 Err(e) => Err(databases::DbError::PoolError(e)) // TODO: Fix this
+                             }
+                        })
+                        .and_then(|pool| Ok(#pool_type(pool)))
+                }
+
+                pub fn fairing() -> impl Fairing {
+                    AdHoc::on_attach(move |rocket| {
+                        if let Some(database_config) = databases::database_config(
+                            #database_name,
+                            &rocket.config().clone()
+                        ) {
+                            return match Self::pool(database_config) {
+                                Ok(pool) => Ok(rocket.manage(pool)),
+                                Err(_) => Err(rocket)
+                            }
+                        } else {
+                            error_!("Could not find database: {}", #database_name);
+                        }
+
+                        Err(rocket)
+                    })
+                }
+            }
+
+            impl Deref for #handler_type {
+                type Target = #(#inner_type)::*;
+
+                fn deref(&self) -> &Self::Target {
+                    &self.0
+                }
+            }
+
+            impl<'a, 'r> FromRequest<'a, 'r> for #handler_type {
+                type Error = ();
+
+                fn from_request(request: &'a Request<'r>) -> request::Outcome<Self, Self::Error> {
+                    let pool = request.guard::<State<#pool_type>>()?;
+
+                    match pool.0.get() {
+                        Ok(conn) => ::rocket::Outcome::Success(#handler_type(conn)),
+                        Err(_) => ::rocket::Outcome::Failure((Status::ServiceUnavailable, ())),
```

Comment:
Did we determine if there was a better error type we could use?

Reply:
I'm not sure. I'll refer back to our previous conversations on this.